src/types.rs

```diff
@@ -0,0 +1,627 @@
+/*
+ * This file is part of Astarte.
+ *
+ * Copyright 2022 SECO Mind Srl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::collections::HashMap;
+use std::time::SystemTime;
+
+use chrono::{DateTime, Utc};
+
+use crate::proto_message_hub::astarte_data_type::Data::{AstarteIndividual, AstarteObject};
+use crate::proto_message_hub::astarte_data_type_individual::IndividualData;
+use crate::proto_message_hub::{
+    AstarteBinaryBlobArray, AstarteBooleanArray, AstarteDataType, AstarteDataTypeIndividual,
+    AstarteDataTypeObject, AstarteDateTimeArray, AstarteDoubleArray, AstarteIntegerArray,
+    AstarteLongIntegerArray, AstarteStringArray,
+};
+
+macro_rules! impl_type_conversion_traits {
+    ( {$( ($typ:ty, $astartedatatype:ident) ,)*}) => {
+
+        $(
+               impl From<$typ> for AstarteDataTypeIndividual {
+                    fn from(d: $typ) -> Self {
+                        AstarteDataTypeIndividual {
+                            individual_data: Some(IndividualData::$astartedatatype(d.into())),
+                            }
+                    }
+                }
+
+                impl From<&$typ> for AstarteDataTypeIndividual {
+                    fn from(d: &$typ) -> Self {
+                        AstarteDataTypeIndividual {
+                            individual_data: Some(IndividualData::$astartedatatype(d.clone().into())),
+                            }
+                    }
+                }
+        )*
+    };
+}
+
+impl_type_conversion_traits!({
+    (f64, AstarteDouble),
+    (i32, AstarteInteger),
+    (bool, AstarteBoolean),
+    (i64, AstarteLongInteger),
+    (&str, AstarteString),
+    (String, AstarteString),
+    (Vec<u8>, AstarteBinaryBlob),
+});
+
+impl From<DateTime<Utc>> for AstarteDataTypeIndividual {
+    fn from(value: DateTime<Utc>) -> Self {
+        let system_time: SystemTime = value.into();
+
+        AstarteDataTypeIndividual {
+            individual_data: Some(IndividualData::AstarteDateTime(system_time.into())),
+        }
+    }
+}
+
+impl From<Vec<f64>> for AstarteDataTypeIndividual {
+    fn from(value: Vec<f64>) -> Self {
```

Comment:
On a similar note, I think in Arrays the argument should be called `values` instead of `value`